analyzer/runtime/extract.go

```diff
@@ -364,15 +364,40 @@ func ExtractRound(blockHeader nodeapi.RuntimeBlockHeader, txrs []nodeapi.Runtime
 					blockTransactionData.Body = body
 					amount = body.Amount.Amount
 					if body.To != nil {
-						// Beware, this is the address of an account in the consensus
-						// layer, not an account in the runtime that generated this event.
+						// This is the address of an account in the consensus
+						// layer, not an account in the runtime that generated this tx.
 						// We do not register it as a preimage.
 						if to, err = uncategorized.StringifySdkAddress(body.To); err != nil {
 							return fmt.Errorf("to: %w", err)
 						}
 					}
 					return nil
 				},
+				ConsensusAccountsDelegate: func(body *consensusaccounts.Delegate) error {
+					blockTransactionData.Body = body
+					amount = body.Amount.Amount
+					// This is the address of an account in the consensus
+					// layer, not an account in the runtime that generated this tx.
+					// We do not register it as a preimage.
+					if to, err = uncategorized.StringifySdkAddress(&body.To); err != nil {
+						return fmt.Errorf("to: %w", err)
+					}
+					return nil
+				},
+				ConsensusAccountsUndelegate: func(body *consensusaccounts.Undelegate) error {
+					blockTransactionData.Body = body
+					// NOTE: The `from` and `to` addresses have swapped semantics compared to most other txs:
+					// Assume R is a runtime address and C is a consensus address. The inverse of Delegate(from=R, to=C) is Undelegate(from=C, to=R).
+					if _, err := registerRelatedSdkAddress(blockTransactionData.RelatedAccountAddresses, &body.From); err != nil {
+						return fmt.Errorf("from: %w", err)
+					}
+					// The `to` is implicitly the account that created this tx. Ref: https://github.com/oasisprotocol/oasis-sdk/blob/eb97a8162f84ae81d11d805e6dceeeb016841c27/runtime-sdk/src/modules/consensus_accounts/mod.rs#L465-L465
+					to = blockTransactionData.SignerData[0].Address
```

Comment:
While the `to` address according to tx-internal semantics is indeed the tx sender (see NOTE comment above), it might make more sense to store `body.From` in our `to` field? The `to` field is internally defined by the indexer (though exposed through the API), with [varying meanings](https://github.com/oasisprotocol/nexus/blob/2ea104969cc4242ca896a7d50bd313024c849432/api/spec/v1.yaml#L2368-L2375) but generally being "the address that is most relevant to this tx, other than the sender".

Reply:
:skull: is this what we do for consensus undelegate?

Reply:
By "consensus undelegate", you mean the `ReclaimEscrow` method from [here](https://github.com/oasisprotocol/oasis-core/blob/6dcc5f29ff908f8c6c5bd3481f20969db009c3be/go/staking/api/api.go#L77-L81), right?
When [inserting consensus txs into the DB](https://github.com/oasisprotocol/nexus/blob/2ea104969cc4242ca896a7d50bd313024c849432/analyzer/consensus/consensus.go#L332-L345), we make no attempt to construct a reasonably generalizable `to` field. So I guess there's no prior art :/

I'm actually inclined to set `to = body.From`. Because otherwise there's no added value in setting the `to` (we're storing `body.To` in the tx sender field already), and so it will be hard to surface undelegations relating to a specific validator.

Edit: Doing so after a slack discussion.